base/multidimensional.jl

```diff
@@ -27,10 +27,11 @@ CartesianIndex{N}(index::NTuple{N,Integer}) = CartesianIndex{N}(index)
 (::Type{CartesianIndex{N}}){N}(index::Integer...) = CartesianIndex{N}(index)
 (::Type{CartesianIndex{N}}){N}() = CartesianIndex{N}(())
 # Un-nest passed CartesianIndexes
-CartesianIndex(index::Union{Integer, CartesianIndex}...) = CartesianIndex(_flatten((), index...))
-_flatten(out) = out
-@inline _flatten(out, i::Integer, I...)        = _flatten((out..., i), I...)
-@inline _flatten(out, i::CartesianIndex, I...) = _flatten((out..., i.I...), I...)
+CartesianIndex(index::Union{Integer, CartesianIndex}...) = CartesianIndex(flatten(index))
+Base.@pure flatten(I) = (_flatten(I...)...,)
```

Comment:
For sure. Come to think of it, I don't need the `@pure` annotation on `_flatten`, just on `flatten`. (right?)

The only reason it was marked pure is because of #17126, but this `flatten` has since been redesigned several times, and in testing with this version I'm getting equal performance with `@inline`. So I'll change it.